message_tagging_service/consumer.py

```diff
@@ -55,3 +53,42 @@ def consume(self, msg):
             # file is empty, catch this case and skip to handle the tag.
             if rule_defs is not None:
                 tagging_service.handle(rule_defs, event_msg)
+
+
+class BuildTagConsumer(fedmsg.consumers.FedmsgConsumer):
+    """Consumer messages on build is tagged
+
+    BuildTagConsumer handles two different format of messages for Fedora and
+    Brew. For Fedora, the message is a simple mapping containing tag name and
+    build info. Whereas, the message sent from Brew contains those information in
+    individual mapping. This consumer handles both of them to get build NVR and
+    tag name.
+    """
+
+    topic = conf.consumer_topics['build-tag-consumer']
+    config_key = 'build-tag-consumer'
+
+    def consume(self, msg):
+        logger.debug('Got message: %r', msg)
+
+        event_msg = msg['body']['msg']
+
+        user = event_msg['user']
+        if isinstance(user, dict):
+            # Handle Brew format.
+            user = user['name']
+            nvr = event_msg['build']['nvr']
+            tag = event_msg['tag']['name']
+        else:
+            # The Fedora case.
+            nvr = '{name}-{version}-{release}'.format(**event_msg)
+            tag = event_msg['tag']
+
+        if user != conf.koji_mts_username:
+            logger.debug('Build %s is not tagged by MTS. Skip to notify.', nvr)
+            return
+
+        messaging.publish('build.tagged', {
```

Comment:
What if tagging fails? I'm not seeing a lot of value in just re-wrapping Koji's message. 

I think it would be very beneficial if MTS published a message when it requested the tagging to be done by Koji. This message should include enough information (task ID) for clients that care about tracking progress.

@mprahl thoughts?

Reply:
> What if tagging fails? I'm not seeing a lot of value in just re-wrapping Koji's message.

I had this concern as well. But, afterwards, I was aware of that this "wrapping" message could simplify clients to handle the tagged module without waiting for Koji/Brew to finish tagging by itself. So, in practice, when a client receives a message from `mts.build.tagged`, it can simply have confidence to handle the tagged module.

However, I agree that it depends. It also works for clients to wait by itself with task ID included in message provided by MTS.

> I think it would be very beneficial if MTS published a message when it requested the tagging to be done by Koji. This message should include enough information (task ID) for clients that care about tracking progress.

Both messages of `buildsys.tag` and `brew.build.tag` do not include task ID of tagging request. It is doable in current implementation in master.

At this moment, I would suggest to not merge this patch until there is real such requirement from customers. Meanwhile, to include tagging request task id inside current `mts.build.tagged` event message. What do you think? :)

Reply:
+1 I like what you suggested very much :)

Reply:
I agree with the sentiment that it's not really that useful to wrap the Koji tag messages. If you think it's that important to notify clients, you could just make the handler that tags the build synchronous (in the main thread or in a separate thread that the main thread doesn't wait for). In other words, it could wait for the tag to be complete and then emit a message. Tagging a build takes a matter of seconds, so you could just have a timeout of like 60 seconds to be safe.